db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBClient.java

```diff
@@ -46,76 +47,97 @@ public static EventStoreDBClient create(EventStoreDBClientSettings settings) {
         if (options == null)
             options = ReadStreamOptions.get();
 
         if (!options.hasUserCredentials())
             options.authenticated(this.credentials);
 
-        return new ReadStream(this.client, streamName, maxCount, options).execute();
+        return new ReadStream(this.client, streamName, maxCount, options).execute(observer);
     }
 
     public CompletableFuture<StreamMetadata> getStreamMetadata(String streamName) {
         return getStreamMetadata(streamName, null);
     }
 
-    public CompletableFuture<StreamMetadata> getStreamMetadata(String streamName, ReadStreamOptions options) {
-
-        return readStream("$$" + streamName, options).thenCompose(result -> {
-            RecordedEvent event = result.getEvents().get(0).getOriginalEvent();
-            CompletableFuture<StreamMetadata> out = new CompletableFuture<>();
+    static class StreamMetadataReadObserver extends ReadObserver<StreamMetadata> {
+        StreamMetadata metadata;
 
+        @Override
+        public void onNext(ResolvedEvent event) {
             try {
-                HashMap<String, Object> source = event.getEventDataAs(HashMap.class);
-
-                out.complete(StreamMetadata.deserialize(source));
-            } catch (Throwable e) {
-                out.completeExceptionally(e);
+                HashMap<String, Object> source = event.getOriginalEvent().getEventDataAs(HashMap.class);
+                metadata = StreamMetadata.deserialize(source);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
             }
+        }
+
+        @Override
+        public void onStreamNotFound() {
+            metadata = new StreamMetadata();
+        }
+
+        @Override
+        public  StreamMetadata onCompleted() {
+            return metadata;
+        }
 
-            return out;
-        }).exceptionally(e -> {
-            if (e.getCause() instanceof StreamNotFoundException) {
-                return new StreamMetadata();
+        @Override
+        public void onError(Throwable error) {
+
+        }
+
+        public StreamMetadata getStreamMetadata() {
+            return metadata;
+        }
+    }
+
+    public CompletableFuture<StreamMetadata> getStreamMetadata(String streamName, ReadStreamOptions options) {
+        final StreamMetadataReadObserver observer = new StreamMetadataReadObserver();
+        return readStream("$$" + streamName, options, observer).exceptionally(e -> {
+            Throwable reason = e.getCause() != null ? e.getCause() : e;
+            if (reason instanceof StreamNotFoundException) {
+                return observer.getStreamMetadata();
             }
 
             throw new RuntimeException(e);
         });
     }
 
-    public CompletableFuture<ReadResult> readAll() {
-        return this.readAll(Long.MAX_VALUE, ReadAllOptions.get());
+    public <A> CompletableFuture<A> readAll(ReadObserver<A> observer) {
```

Comment:
**SUGGESTION:** I think we should also add an overload without the observer and set the `new EventCollectorReadObserver()` for that case. It will be more accessible for the people using classically to get all events from the short stream.